### ISSUE TITLE
International Addresses Paritally Imported (Feature)

### ISSUE THREAD

**Issue by mphillips12** [laralee3/restaurantAdder#7]
**ISSUE:** 

When adding an international address, only part of the details are added to the worksheet

**STEPS TO REPRO:**

1. Navigate to Restaurant Adder
2. Type "Mr. Young Men" in the search bar
3. Add Mr. Young Men
4. Navigate to Google Sheets

**ACTUAL:**

No street address is added; zip code is incomplete

**EXPECTED:** 

Location details are complete

**ADDITIONAL INFO:**

1. With Canadian addresses, the address imports fine, but the zip code field is blank (check Tim Hortons)
2. But with Italian addresses everything imports as expected (check Osteria Francescana)
3. Suspect this is an issue with addresses written in Japanese

**Comment by laralee3** [comment]
Need to figure out and implement whatever the common standard is that would work across all addresses. 

For a look at how the data is coming back, see here in the "Address Types and Address Component Types" section: https://developers.google.com/maps/documentation/geocoding/intro

One of the alternatives I would suggest: the places api call we're making also returns a "formatted" address as a single string. Since we're using lat/lng for marker placement in tableau anyway (is that correct?), **we should reduce the address columns to a single column containing a formatted address string** and skip parsing all the address components. Otherwise we'd have to figure out how to find a standard for every, single, country that we might want to go to a restaurant in, and that sounds like a massively bigger time investment than a project this scale should get.

Let me know if the assumption about Tableau marker placement is incorrect @mphillips12

**Comment by mphillips12** [comment]
Yeah, I'm using lat/long for the mapping, and City/State for some other cosmetic stuff. If we put it all in one column I'd have to do some string parsing on the Tableau side, but it shouldn't be too difficult as long as we have a standard delimiter like commas or something.

**Comment by laralee3** [comment]
Right, that does reduce the filtering in tableau...maybe a combination? We can pull states/countries/neighborhoods out from address components, but utilize a full "address" string from the formatted_address as well like "1600 Amphitheatre Parkway, Mountain View, CA 94043, USA". No reason we can't do both.

**Comment by mphillips12** [comment]
Yeah, how difficult would that be on the code side? I say we should go that route so I don't have to fix everything with calculations on the Tableau side, lol

**Comment by laralee3** [comment]
Not hard. Just let me know all the different columns you want for filtering. 

-State
-Country
-Neighborhood?

**Comment by mphillips12** [comment]
State and Country should be fine for now. Anything else I can parse out from the formatted address field.

**Comment by laralee3** [comment]
@mphillips12 so I can implement some changes this weekend; already have some updates. Can you set the sheet to the columns you wanted that this would write to? Example below

- name 
- cuisine type
- address (formatted, such as `1950 6th Ave, Seattle, WA 98101, USA`)
- State
- Country
- Website
- Google Rating
- lat
- lng

I'll look at implemented this to be more dynamic so you don't have to worry about order and such

**Comment by mphillips12** [comment]
@laralee3 okay so the new format is on the Restaurants_Formatted tab with the new columns in blue. When I was doing this I realized another feature request I'll have to log haha. But anyway, if you make the changes to the Restaurants Formatted sheet for now, I can replace the data source in Tableau later when I go to update it, and this way the current view won't be broken.

**Comment by laralee3** [comment]
New build is up: https://laralee3.github.io/restaurantAdder

It's now targeting the Restaurants_Formatted sheet, so new restaurants will be added there.

We might need to change the name of the "state" column, as it doesn't make sense for international addresses. For example, try adding the following restaurants:
- Jiro's Dining (the one in Japan)
- Restaurant Gordon Ramsay (England)
- Lung King Heen (Hong Kong)

The "state" for each result is: 東京都, England, and Hong Kong Island. Google is returning this as "administrative_area_level_1", but we should be able to come up with a more appealing way of titling this column...

If the column name is changed let me know, as it'll break; the translation between google names and our names has to be hard coded as a matter of course, so we should decide on it once and set it as final. 

Changing sheet name will break it too, but just let me know and I can update.

**Comment by mphillips12** [comment]
@laralee3 Most international data sets will list the field as **"State/Province"** so I can change the column name to that. The administrative level 1 thing is weird (like using country code "GB" and not "UK" when GB is basically outdated), so I'll just have to do some creative clean up on the Tableau side to make it look nicer. 

More info on the standard for admin level 1: https://en.wikipedia.org/wiki/ISO_3166-2

**Comment by lara-glympse** [comment]
Cool, I'll update to point at "State/Province"

**Comment by mphillips12** [comment]
@laralee3  New entries in the address_formatted field are getting imported as "address, city, state, zip, **country**"

Is that last country entry intended/necessary? (I ask because it's resulting in having to do some additional formatting/cleanup in tableau, haha)

**Comment by laralee3** [comment]
address_formatted is the entirety of what google is giving us. I can filter it out if you want, but I don't know how consistent that data is returned. Guessing 100% in this case, so it should be safe. Let me know if you want it gone

**Comment by mphillips12** [comment]
Hm, okay. For some reason I thought it wasn't tacking on the country part initially. If it's too much of a hassle to filter it out, I can probably do some string calcs in Tableau to ignore it. But yeah, filter it out if you can haha

**Comment by laralee3** [comment]
@mphillips12 Updated, try now

**Comment by mphillips12** [comment]
Perfect! Awesome!

**Comment by mphillips12** [comment]
@laralee3 i noticed today that google for some reason has two different ways it handles Japanese addresses (attached screenshots). Not sure if it's feasible to distinguish between the two from a code perspective. What do you think? 

![version 1](https://user-images.githubusercontent.com/28489499/26883060-f878feba-4b50-11e7-82c6-3448cfb59269.jpg)
![version 2](https://user-images.githubusercontent.com/28489499/26883059-f85ef5ba-4b50-11e7-9398-80964da0662e.jpg)
